File: vciplreportsv1001/vcipl1001/report/top_most_selling_items/top_most_selling_items.py
import frappe
from datetime import date


def execute(filters=None):
    filters = frappe._dict(filters or {})

    # ---------- DEFAULT FINANCIAL YEAR ----------
    if not filters.get("from_date") or not filters.get("to_date"):
        today = date.today()
        fy_year = today.year if today.month > 3 else today.year - 1

        filters.from_date = date(fy_year, 4, 1)
        filters.to_date = date(fy_year + 1, 3, 31)

    return get_columns(), get_data(filters)


# -------------------- COLUMNS --------------------
def get_columns():
    return [
        {
            "label": "Item Code",
            "fieldname": "item_code",
            "fieldtype": "Link",
            "options": "Item",
            "width": 150
        },
        {
            "label": "Item Name",
            "fieldname": "item_name",
            "fieldtype": "Data",
            "width": 200
        },
        {
            "label": "Item Group",
            "fieldname": "item_group",
            "fieldtype": "Link",
            "options": "Item Group",
            "width": 150
        },
        {
            "label": "Total Sales Amount",
            "fieldname": "total_amount",
            "fieldtype": "Currency",
            "width": 150
        },
        {
            "label": "Total Sold Qty",
            "fieldname": "total_qty",
            "fieldtype": "Float",
            "width": 120
        },
        {
            "label": "Total Stock Qty",
            "fieldname": "total_stock_qty",
            "fieldtype": "Float",
            "width": 140
        }
    ]


# -------------------- DATA --------------------
def get_data(filters):
    limit = int(filters.get("record_limit") or 50)

    params = {
        "docstatus": 1,
        "from_date": filters.from_date,
        "to_date": filters.to_date,
        "item_type": filters.get("custom_item_type")
    }

    # ---------------- SALES QUERY ----------------
    sales_query = f"""
        SELECT
            sii.item_code,
            i.item_name,
            i.item_group,
            SUM(sii.amount) AS total_amount,
            SUM(sii.qty) AS total_qty
        FROM `tabSales Invoice Item` sii
        INNER JOIN `tabSales Invoice` si ON si.name = sii.parent
        INNER JOIN `tabItem` i ON i.name = sii.item_code
        WHERE
            si.docstatus = %(docstatus)s
            AND si.posting_date BETWEEN %(from_date)s AND %(to_date)s
            AND (%(item_type)s IS NULL OR i.custom_item_type = %(item_type)s)
        GROUP BY sii.item_code, i.item_name, i.item_group
        ORDER BY total_amount DESC
        LIMIT {limit}
    """

    rows = frappe.db.sql(sales_query, params, as_dict=True)

    if not rows:
        return []

    item_codes = [r["item_code"] for r in rows]

    # ---------------- STOCK ----------------
    bins = frappe.get_all(
        "Bin",
        filters={"item_code": ["in", item_codes]},
        fields=["item_code", "actual_qty"]
    )

    stock_map = {}
    for b in bins:
        stock_map[b.item_code] = stock_map.get(b.item_code, 0) + b.actual_qty

    for r in rows:
        r["total_stock_qty"] = stock_map.get(r["item_code"], 0)

    return rows

File: vciplreportsv1001/vcipl1001/report/top_most_selling_items/top_most_selling_items.js
frappe.query_reports["Top Most Selling Items"] = {

    onload: function (report) {
        const today = frappe.datetime.get_today();
        const today_obj = frappe.datetime.str_to_obj(today);

        const year = today_obj.getMonth() > 2
            ? today_obj.getFullYear()
            : today_obj.getFullYear() - 1;

        // ✅ Default values
        report.set_filter_value("custom_item_type", "Finished Goods");
        report.set_filter_value("record_limit", 50);
        report.set_filter_value("from_date", `${year}-04-01`);
        report.set_filter_value("to_date", `${year + 1}-03-31`);
        report.set_filter_value("show_all_items", 0);
    },

    // ✅ FINAL FIX → Show only Item Code (no name)
    formatter: function (value, row, column, data, default_formatter) {

        // Apply default formatter first
        value = default_formatter(value, row, column, data);

        // 🔥 Override Item Code display
        if (column.fieldname === "item_code" && data) {
            return data.item_code || "";
        }

        return value;
    },

    filters: [
        {
            fieldname: "custom_item_type",
            label: __("Item Type"),
            fieldtype: "Data",
            default: "Finished Goods"
        },
        {
            fieldname: "from_date",
            label: __("From Date"),
            fieldtype: "Date",
            reqd: 1
        },
        {
            fieldname: "to_date",
            label: __("To Date"),
            fieldtype: "Date",
            reqd: 1
        },
        {
            fieldname: "show_all_items",
            label: __("Show All Items"),
            fieldtype: "Check",
            default: 0
        },
        {
            fieldname: "record_limit",
            label: __("Top Records"),
            fieldtype: "Select",
            options: [
                { label: "Top 50", value: 50 },
                { label: "Top 100", value: 100 }
            ],
            default: 50,
            reqd: 1,
            depends_on: "eval:!doc.show_all_items"
        }
    ]
};